src/docs/rts-new.md

```diff
@@ -0,0 +1,170 @@
+
+# Interaction with transaction IDs
+
+This is a little simpler than the timestamp rules - we use normal snapshot
+visibility in most of the places we look at timestamps (see the timestamps
+section). One exception is that we treat all data as visible when RTS is
+called from the API (as opposed to startup or shutdown). 
+
+We also clear the transaction ID of updates we look at during RTS, because
+the connection's write generation will be initialised after RTS and the
+updates in the cache would be problematic for other parts of the code if
+they had a "legitimate" transaction ID.
+
+# Operations on the history store
+
+RTS touches the history store for a lot of things, primarily for deleting,
+but it might also need to read an update to move it to the data store.
+
+The deletion code is called from the general update aborting code, where we
+might need to delete the first stable update we found when it came from the
+history store. There's also a somewhat subtle case where a new key can be
+added to a page, and then the page checkpointed, leading to a situation
+where we have updates to the key in the history store but not the disk
+image.
+
+History store deletions are done using a standard history store cursor,
+starting at the highest possible start timestamp, and removing keys from
+newest to oldest until we hit an update with a start timestamp larger than
+the rollback timestamp.
+
+Note: in-memory databases don't have a history store, so none of this
+section will apply.
+
+# Iterating over B-Trees
+
+This is one of the simplest parts of the RTS subsystem, as a metadata cursor
+is specifically designed for these sorts of "iterate over a B-Tree"
+operations. We use this functionality for RTS, with the caveat that we tweak
+a few options so that corrupt or missing files don't raise errors. The
+reasoning here is that RTS is orthogonal to salvage, and might be called as
+part of salvage, so we simply need to "do our best". We do log the issues we
+find, however.
+
+Less obviously, the history store needs one final inspection regardless of
+whether any other B-Trees were touched. This is because we usually rely on
+getting to the history store "via" the data store, but if we don't touch a
+B-Tree, that doesn't mean there aren't updates in the history store after
+the stable timestamp.
+
+# Iterating over pages
+
+RTS tries to avoid looking at a page if it can. There are two ways we can
+avoid instantiating a page: if it was truncated at/before the stable
+timestamp, and if the page state is anything other than on-disk. If neither
+of these are satisfied, we need to instantiate the page to look at it, but
+that doesn't mean we look at the contents. Rather, we use the existing tree
+walk custom skipping functionality to check the page metadata, stuff like
+reconciliation results. These checks are fairly low-level, so they're
+outside the scope of this document.
+
+One final note is that we deliberately do not call RTS for internal pages.
```

Comment:
@kommiharibabu I suspect this is because they won't have "updates" to roll back as such, and we handle tree maintenance (node merging, splitting etc) during reconciliation. Is this correct? If so, I'll add a short explanation.

Reply:
Yes, we do not touch the internal pages as they do not have updates. Just to be clear, we read the internal pages into the cache to abort any unstable updates on the leaf pages of the internal pages.

Reply:
Thanks for confirming, I've added some details to this effect in ae692da.